cpp/bench/sg/kmeans.cu

```diff
@@ -0,0 +1,122 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuML.hpp>
+#include <kmeans/kmeans.hpp>
+#include <utility>
+#include "benchmark.cuh"
+
+namespace ML {
+namespace Bench {
+namespace kmeans {
+
+struct Params {
+  DatasetParams data;
+  BlobsParams blobs;
+  ML::kmeans::KMeansParams kmeans;
+};
+
+template <typename D>
+class KMeans : public BlobsFixture<D> {
+ public:
+  KMeans(const std::string& name, const Params& p)
+    : BlobsFixture<D>(p.data, p.blobs), kParams(p.kmeans) {
+    this->SetName(name.c_str());
+  }
+
+ protected:
+  void runBenchmark(::benchmark::State& state) override {
+    if (!this->params.rowMajor) {
+      state.SkipWithError("KMeans only supports row-major inputs");
+    }
+    auto& handle = *this->handle;
+    auto stream = handle.getStream();
+    for (auto _ : state) {
+      CudaEventTimer timer(handle, state, true, stream);
+      ML::kmeans::fit_predict(handle, kParams, this->data.X, this->params.nrows,
+                              this->params.ncols, centroids, labels, inertia,
+                              nIter);
+    }
+  }
+
+  void allocateBuffers(const ::benchmark::State& state) override {
+    auto allocator = this->handle->getDeviceAllocator();
+    auto stream = this->handle->getStream();
+    labels =
+      (int*)allocator->allocate(this->params.nrows * sizeof(int), stream);
+    centroids = (D*)allocator->allocate(
+      this->params.nclasses * this->params.ncols * sizeof(D), stream);
+  }
+
+  void deallocateBuffers(const ::benchmark::State& state) override {
+    auto allocator = this->handle->getDeviceAllocator();
+    auto stream = this->handle->getStream();
+    allocator->deallocate(
+      centroids, this->params.nclasses * this->params.ncols * sizeof(D),
+      stream);
+    allocator->deallocate(labels, this->params.nrows * sizeof(int), stream);
+  }
+
+ private:
+  ML::kmeans::KMeansParams kParams;
+  int* labels;
+  D* centroids;
+  D inertia;
+  int nIter;
+};
+
+std::vector<Params> getInputs() {
+  std::vector<Params> out;
+  Params p;
+  p.data.rowMajor = true;
+  p.blobs.cluster_std = 1.0;
+  p.blobs.shuffle = false;
+  p.blobs.center_box_min = -10.0;
+  p.blobs.center_box_max = 10.0;
+  p.blobs.seed = 12345ULL;
+  p.kmeans.init = ML::kmeans::KMeansParams::InitMethod(0);
+  p.kmeans.max_iter = 300;
+  p.kmeans.tol = 1e-4;
+  p.kmeans.verbose = false;
+  p.kmeans.seed = int(p.blobs.seed);
+  p.kmeans.metric = 0;  // L2
+  p.kmeans.inertia_check = true;
+  std::vector<std::pair<int, int>> rowcols = {
+    {160000, 64},
```

Comment:
We'll want to benchmark both skinny and fat as well, right?

Reply:
fair enough. Let me add some more.